draco/analysis/transform.py

```diff
@@ -889,3 +892,159 @@ def process(self, polcont):
+        ap_orig = self.telescope._angpos
+        self.telescope._angpos = ap_ref
+        beam_ref = self.telescope.beam(feed, freq)
+        self.telescope._angpos = ap_orig
+
+        # Normalise omega by the squared magnitude of the beam at the reference position
+        beam_ref = np.sum(np.abs(beam_ref) ** 2)
+        omega /= beam_ref
+
+        return omega
+
+    def process(self, sstream: containers.SiderealStream) -> containers.SiderealStream:
+        """Apply the brightness temperature to flux conversion to the data.
+
+        Parameters
+        ----------
+        sstream
+            The visibilities to apply the conversion to. They are converted to/from
+            brightness temperature units depending on the setting of `convert_Jy_to_K`.
+
+        Returns
+        -------
+        new_sstream
+            Visibilities with the conversion applied. This may be the same as the input
+            container if `share == "all"`.
+        """
+
+        import scipy.constants as c
+
+        sstream.redistribute("freq")
+
+        # Get the local frequencies in the sidereal stream
+        sfreq = sstream.vis.local_offset[0]
+        efreq = sfreq + sstream.vis.local_shape[0]
+        local_freq = sstream.freq[sfreq:efreq]
+
+        # Get the indices of the incoming frequencies as far as the telescope class is
+        # concerned
+        local_freq_inds = []
+        for freq in local_freq:
+            local_freq_inds.append(np.argmin(np.abs(self.telescope.frequencies - freq)))
+
+        # Get the feedpairs we have data for and their beamclass (usually this maps to
+        # polarisation)
+        feedpairs = structured_to_unstructured(sstream.prodstack)
+        beamclass_pairs = self.telescope.beamclass[feedpairs]
+
+        # Calculate all the unique beams that we need to calculate areas for
+        unique_beamclass, bc_index = np.unique(beamclass_pairs, return_index=True)
+
+        # Calculate any missing beam areas and to the cache
+        for beamclass, bc_ind in zip(unique_beamclass, bc_index):
+            feed_ind = feedpairs.ravel()[bc_ind]
+
+            for freq, freq_ind in zip(local_freq, local_freq_inds):
+                key = (beamclass, freq)
+
+                if key not in self._omega_cache:
+                    self._omega_cache[key] = self._beam_area(feed_ind, freq_ind)
+
+        # Loop over all frequencies and visibilities and get the effective primary
+        # beam area for each
+        om_ij = np.zeros((len(local_freq), sstream.vis.shape[1]))
+        for fi, freq in enumerate(local_freq):
+            for bi, (bci, bcj) in enumerate(beamclass_pairs):
+                om_i = self._omega_cache[(bci, freq)]
+                om_j = self._omega_cache[(bcj, freq)]
+                om_ij[fi, bi] = (om_i * om_j) ** 0.5
+
+        # Calculate the Jy to K conversion
+        wavelength = c.c / (local_freq * 10 ** 6)
+        Jy_to_K = (
```

Comment:
For external beam models that are derived from the data, there may be missing frequencies where the beam is set to zero.  In that case, the beam solid angle will be zero and we may want to use `tools.invert_no_zero` here and elsewhere to prevent NaNs.